firmware/shared/code/I2C.c

```diff
@@ -0,0 +1,311 @@
+/*
+ * I2C.c
+ *
+ *  Created on: Oct 10, 2020
+ *      Author: Shuhang Tan
+ *      Reference: https://github.com/uvic-auvic/Firmware-Motor_Controller_2.0/blob/master/src/I2C.c
+ *                 (I2C code for motor controller 2)
+ */
+
+#include "I2C.h"
+#include "assert.h"
+#include "RCCHelper.h"
+#include <string.h>
+#include <stdlib.h>
+#include "LED.h"
+
+// I2C SDA and SCL pins
+#define _I2C_SDA_GPIO  (GPIO_Pin_9) // PC9
+#define _I2C_SCL_GPIO  (GPIO_Pin_8) // PA8
+
+// I2C sendBuffer size
+#define _I2C_BUFFER_SIZE    (10U)
+
+typedef struct
+{
+	uint8_t Buffer[_I2C_BUFFER_SIZE];
+} I2C_data_S;
+
+// Adding an idle state because reading the BUSY bit in SR2 can cause problems
+typedef enum{
+	I2C_STATE_SEND,    // Write
+	I2C_STATE_RECEIVE, // Read
+	I2C_STATE_IDLE,    // Idle
+}I2C_state_S;
+
+// Global variables and their initialization
+extern const I2C_config_S I2C_config;
+
+static I2C_data_S I2C_send_data;
+static I2C_data_S I2C_receive_data;
+static I2C_state_S I2C_state = I2C_STATE_IDLE;
+
+uint8_t slave_address = 0x0;
+
+// Length of the data to be sent or received (in bytes)
+volatile uint8_t data_length = 0;
+
+// Pointer to the buffers
+uint8_t* sendBufferPtr;
+uint8_t* receiveBufferPtr;
+
+void I2C_init(void)
+{
+	// Check that all configs are valid
+	assert(IS_GPIO_PIN_SOURCE(I2C_config.HWConfig->SDAPin));
+	assert(IS_GPIO_ALL_PERIPH(I2C_config.HWConfig->SDAPort));
+	assert(IS_GPIO_PIN_SOURCE(I2C_config.HWConfig->SCLPin));
+	assert(IS_GPIO_ALL_PERIPH(I2C_config.HWConfig->SCLPort));
+	assert(IS_I2C_ALL_PERIPH(I2C_config.HWConfig->I2CPeriph));
+
+	// GPIO Init
+	RCCHelper_clockCmd(I2C_config.HWConfig->SDAPort, ENABLE);
+	if (I2C_config.HWConfig->SDAPort != I2C_config.HWConfig->SCLPort)
+	{
+		RCCHelper_clockCmd(I2C_config.HWConfig->SCLPort, ENABLE);
+	}
+
+	GPIO_InitTypeDef GPIO_InitStruct;
+	GPIO_InitStruct.GPIO_Pin   = I2C_config.HWConfig->SDAPin;
+	GPIO_InitStruct.GPIO_Mode  = GPIO_Mode_AF;
+	GPIO_InitStruct.GPIO_Speed = GPIO_Speed_100MHz;
+	GPIO_InitStruct.GPIO_OType = GPIO_OType_OD;
+	GPIO_InitStruct.GPIO_PuPd  = GPIO_PuPd_UP;
+	GPIO_Init(I2C_config.HWConfig->SDAPort, &GPIO_InitStruct);
+
+	GPIO_InitStruct.GPIO_Pin   = I2C_config.HWConfig->SCLPin;
+	GPIO_InitStruct.GPIO_Mode  = GPIO_Mode_AF;
+	GPIO_InitStruct.GPIO_Speed = GPIO_Speed_100MHz;
+	GPIO_InitStruct.GPIO_OType = GPIO_OType_OD;
+	GPIO_InitStruct.GPIO_PuPd  = GPIO_PuPd_UP;
+	GPIO_Init(I2C_config.HWConfig->SCLPort, &GPIO_InitStruct);
+
+	// Alternative function config for SDA and SCL
+	GPIO_PinAFConfig(I2C_config.HWConfig->SDAPort, I2C_config.HWConfig->SDAPin, I2C_config.HWConfig->I2CAFNum);

```

Comment:
Argument 2 should be the GPIO number, not the GPIO mask.
Should be one of `GPIO_PinSource_X`, not `GPIO_Pin_X`.
I'll explain this to you more on Tuesday.